source/agora/consensus/EnrollmentPool.d

```diff
@@ -0,0 +1,374 @@
+    on the network to be a validator, which means the enrollment information
+    needs to be confirmed on the consensus protocol to be a validator. The
+    information is stored in a table of SQLite.
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.consensus.EnrollmentPool;
+
+import agora.common.crypto.ECC;
+import agora.common.crypto.Key;
+import agora.common.crypto.Schnorr;
+import agora.common.Hash;
+import agora.common.Serializer;
+import agora.consensus.data.Block;
+import agora.consensus.data.Enrollment;
+import agora.consensus.data.PreImageInfo;
+import agora.consensus.data.UTXOSet;
+import agora.consensus.Validation;
+import agora.utils.Log;
+
+import d2sqlite3.database;
+import d2sqlite3.library;
+import d2sqlite3.results;
+import d2sqlite3.sqlite3;
+
+mixin AddLogger!();
+
+/// Ditto
+public class EnrollmentPool
+{
+    /// SQLite DB instance
+    private Database db;
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            db_path = path to the database file, or in-memory storage if
+                        :memory: was passed
+
+    ***************************************************************************/
+
+    public this (string db_path)
+    {
+        this.db = Database(db_path);
+
+        // create the table for enrollment pool if it doesn't exist yet
+        this.db.execute("CREATE TABLE IF NOT EXISTS enrollment_pool " ~
+            "(key TEXT PRIMARY KEY, val BLOB NOT NULL)");
+    }
+
+    /***************************************************************************
+
+        Shut down the database
+
+        Note: this method must be called explicitly, and not inside of
+        a destructor.
+
+    ***************************************************************************/
+
+    public void shutdown ()
+    {
+        this.db.close();
+    }
+
+    /***************************************************************************
+
+        Add a enrollment data to the enrollment pool
+
+        Params:
+            block_height = the current block height in the ledger
+            finder = the delegate to find UTXOs with
+            enroll = the enrollment data to add
+
+        Returns:
+            true if the enrollment data has been added to the enrollment pool
+
+    ***************************************************************************/
+
+    public bool add (ulong block_height, scope UTXOFinder finder,
+        const ref Enrollment enroll) @safe nothrow
+    {
+        static ubyte[] buffer;
+
+        // check validity of the enrollment data
+        if (auto reason = isInvalidEnrollmentReason(enroll, block_height + 1,
+            finder))
```

Comment:
The `block_height` is not used properly, the function of `isInvalidEnrollmentReason` should be changed.